pkg/metrics/collector/netlink_metrics.go

```diff
@@ -15,29 +15,156 @@ package collector
 
 import (
 	"fmt"
+	"sync"
 	"syscall"
+	"time"
 
 	"github.com/GoogleCloudPlatform/netd/pkg/tcp_metrics/inetdiag"
 	"github.com/GoogleCloudPlatform/netd/pkg/tcp_metrics/parser"
+	"github.com/GoogleCloudPlatform/netd/pkg/tcp_metrics/tcp"
 
+	"github.com/golang/glog"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/vishvananda/netlink/nl"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
 )
 
 var (
 	TCPConnectionsDesc = prometheus.NewDesc(
-		"tcp_connections",
+		"active_tcp_connections",
 		"Current amount of tcp connections on a node",
-		nil, nil,
+		[]string{"pod_name", "namespace_name"}, nil,
 	)
 
 	ActiveTCPRetransmits = prometheus.NewDesc(
 		"active_tcp_retransmits",
 		"Current amount of tcp retransmits on a node",
-		nil, nil,
+		[]string{"pod_name", "namespace_name"}, nil,
 	)
 )
 
+type netlinkCollector struct {
+}
+
+type netlinkStats struct {
+	retransmits    uint64
+	tcpConnections uint64
+}
+
+type podDef struct {
+	name      string
+	namespace string
+}
+
+var (
+	mapMux sync.RWMutex
+	ipMap  = make(map[string]podDef)
+)
+
+func init() {
+	glog.Infof("Initializing netlink metrics")
+	err := createPodWatch()
+	if err != nil {
+		glog.Infof("Error, pod watch not running with err, %q", err)
+	}
+	registerCollector("netlink_metrics", NewNetlinkCollector)
+}
+
+// Functionality for multithreaded map usage
+func safeIPWrite(key string, value podDef) {
+	mapMux.Lock()
+	defer mapMux.Unlock()
+	ipMap[key] = value
+}
+
+func safeIPDelete(key string) {
+	mapMux.Lock()
+	defer mapMux.Unlock()
+	delete(ipMap, key)
+}
+
+func safeIPRead(key string) (podDef, bool) {
+	mapMux.RLock()
+	defer mapMux.RUnlock()
+	podDef, ok := ipMap[key]
+	return podDef, ok
+}
+
+// Functionality for pod watching
+func createPodWatch() error {
+	// creates the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return err
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	kubeInformerFactory := informers.NewSharedInformerFactory(clientset, time.Second*30)
+	podInformer := kubeInformerFactory.Core().V1().Pods().Informer()
```

Comment:
To restrict this to just pods on the current node, add a field selector with spec.nodeName=<node_name>

See for example: https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/pod.go#L123

(They use ParseSelectorOrDie, which is a bit extreme for us here. No need to die here, this watch is not critical to the operation of netd. Use one of the other functions; like OneTermEqualSelector in  https://github.com/kubernetes/apimachinery/blob/master/pkg/fields/selector.go#L465

Reply:
Done